plugins/modules/adguardhome_rewrite.py

```diff
@@ -0,0 +1,185 @@
+      type: str
+      sample: 192.168.178.71
+    domain:
+      description: Domain of the rewrite.
+      type: str
+      sample: dns.osuv.de
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.basic import missing_required_lib
+from ansible_collections.community.dns.plugins.module_utils.adguardhome.api import (
+    AdGuardHomeAPIHandler
+)
+import traceback
+
+YAML_IMP_ERR = None
+try:
+    import yaml
+    HAS_YAML = True
+except Exception:
+    YAML_IMP_ERR = traceback.format_exc()
+    HAS_YAML = False
+
+
+def find_and_compare(rules, domain, answer):
+    domain_exists = False
+    value_is_different = False
+    target = {}
+    for rule in rules:
+        if rule["domain"] == domain:
+            domain_exists = True
+            target = {"domain": domain, "answer": rule["answer"]}
+            if rule["answer"] != answer:
+                value_is_different = True
+            break
+    return domain_exists, value_is_different, target
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            username=dict(type='str', required=True),
+            password=dict(type='str', required=True, no_log=True),
+            host=dict(type='str', required=True),
+            ssl_verify=dict(type='bool', default=True),
+            state=dict(type='str', default='present', choices=['present', 'absent']),
+            answer=dict(type='str', required=False),
+            domain=dict(type='str', required=True),
+        ),
+        supports_check_mode=True,
+        required_if=[['state', 'present', ['answer']]],
+    )
+
+    if not HAS_YAML:
+        module.fail_json(
+            msg=missing_required_lib("pyyaml", url='https://pyyaml.org/wiki/PyYAMLDocumentation'),
+            exception=YAML_IMP_ERR
+        )
+
+    domain = module.params.get('domain')
+    answer = module.params.get('answer')
+    state = module.params.get('state')
+
+    adguardhome = AdGuardHomeAPIHandler(module.params, module.fail_json)
+
+    before = adguardhome.list()
+    changed = False
+
+    DOMAIN_EXISTS, VALUE_IS_DIFFERENT, TARGET = find_and_compare(before, domain, answer)
+    if state == 'present':
+        if not DOMAIN_EXISTS and not module.check_mode and not VALUE_IS_DIFFERENT:
+            changed = True
+            if module.check_mode:
+                checked_mode_after = before + [{"answer": answer, "domain": domain}]
+            else:
+                adguardhome.add_or_delete(domain, answer, "add", TARGET)
+
+        if DOMAIN_EXISTS and VALUE_IS_DIFFERENT:
+            changed = True
+            adguardhome.update(domain, answer, TARGET)
+
+    else:
+        if DOMAIN_EXISTS and not module.check_mode:
+            adguardhome.add_or_delete(domain, answer, "delete", TARGET)
+            changed = True
+
+    after = adguardhome.list()
+
+    diff_item = dict(
+        before=yaml.safe_dump(before),
+        after=yaml.safe_dump(after)
+    )
```

Comment:
Do I understand it correctly that you need PyYAML only to format the diffs? Why do you format diffs at all, and not let the user's configured callback plugin do that? If the problem is that it's a list, simply embed the list into a dictionary with a single key/value pair.